MarkupsToModel/MRML/vtkMRMLMarkupsToModelNode.cxx

```diff
@@ -267,165 +248,318 @@ void vtkMRMLMarkupsToModelNode::SetAndObserveMarkupsNodeID( const char* markupsI
   vtkNew<vtkIntArray> events;
   events->InsertNextValue( vtkCommand::ModifiedEvent );
   events->InsertNextValue( vtkMRMLMarkupsNode::PointModifiedEvent ); // PointEndInteractionEvent 
-  this->SetAndObserveNodeReferenceID( MARKUPS_ROLE, markupsId, events.GetPointer() );
+  this->SetAndObserveNodeReferenceID( INPUT_MARKUPS_ROLE, markupsId, events.GetPointer() );
   this->InvokeCustomModifiedEvent(InputDataModifiedEvent);
 }
 
+void vtkMRMLMarkupsToModelNode::SetAndObserveModelNodeID( const char* modelId )
+{
+  // SetAndObserveNodeReferenceID does not handle nicely setting of the same
+  // node (it should simply ignore the request, but it adds another observer instead)
+  // so check for node equality here.
+  const char* currentNodeId = this->GetNodeReferenceID( OUTPUT_MODEL_ROLE );
+  if ( modelId != NULL && currentNodeId != NULL )
+  {
+    if ( strcmp( modelId, currentNodeId ) == 0 )
+    {
+      // not changed
+      return;
+    }
+  }
+  this->SetAndObserveNodeReferenceID( OUTPUT_MODEL_ROLE, modelId );
+}
+
 void vtkMRMLMarkupsToModelNode::ProcessMRMLEvents( vtkObject *caller, unsigned long event, void *callData )
 {
   vtkMRMLNode* callerNode = vtkMRMLNode::SafeDownCast( caller );
   if ( callerNode == NULL ) return;
 
-  if (this->GetMarkupsNode() && this->GetMarkupsNode()==caller)
+  if ( this->GetMarkupsNode() && this->GetMarkupsNode()==caller )
   {
     this->InvokeCustomModifiedEvent(InputDataModifiedEvent);
   }
 }
 
-void vtkMRMLMarkupsToModelNode::RemoveAllMarkups()
+std::string vtkMRMLMarkupsToModelNode::GetModelNodeName()
 {
-  vtkMRMLMarkupsFiducialNode* markupsNode = vtkMRMLMarkupsFiducialNode::SafeDownCast( this->GetNodeReference( MARKUPS_ROLE ) );
-  markupsNode->RemoveAllMarkups();
+  vtkMRMLModelNode* modelNode = vtkMRMLModelNode::SafeDownCast( this->GetNodeReference( OUTPUT_MODEL_ROLE ) );
+  if ( modelNode == NULL )
+  {
+    return std::string( this->GetID() ).append( "Model" );
+  }
+  else
+  {
+    return std::string( modelNode->GetName() );
+  }
 }
 
-void vtkMRMLMarkupsToModelNode::RemoveLastMarkup()
+std::string vtkMRMLMarkupsToModelNode::GetModelDisplayNodeName()
```

Comment:
The logic class seems to use it a few times, so I made it a static logic function